consensus/src/experimental/buffer_manager.rs

```diff
@@ -172,38 +174,19 @@ impl StateManager {
     /// check if the items at and after the execution root is already executed
     /// if yes, move the execution root to the first *unexecuted* item.
     /// if there is no such item, set it to none.
-    fn try_advance_executed_root(&mut self) {
-        let mut cursor = self.execution_root.clone();
-        while cursor.is_some() {
-            {
-                let buffer_item = get_elem(&cursor);
-                if !matches!(&*buffer_item, BufferItem::Executed(_)) {
-                    break;
-                }
-            }
-            cursor = get_next(&cursor);
-        }
-        self.execution_root = cursor;
+    fn advance_executed_root(&mut self) {
+        let cursor = self.execution_root.clone();
+        self.execution_root = find_elem(cursor, |item| !matches!(*item, BufferItem::Executed(_)));
     }
 
     /// check if the items at and after the signing root is already signed
     /// if yes, move the signing root to the first *unsigned* item.
     /// if there is no such item, set it to none.
-    fn try_advance_signing_root(&mut self) {
-        let mut cursor = self.signing_root.clone();
-        while cursor.is_some() {
-            {
-                let buffer_item = get_elem(&cursor);
-                if !matches!(
-                    &*buffer_item,
-                    BufferItem::Signed(_) | BufferItem::Aggregated(_)
-                ) {
-                    break;
-                }
-            }
-            cursor = get_next(&cursor);
-        }
-        self.signing_root = cursor;
+    fn advance_signing_root(&mut self) {
+        let cursor = self.signing_root.clone();
+        self.signing_root = find_elem(cursor, |item| {
+            !matches!(*item, BufferItem::Signed(_) | BufferItem::Aggregated(_))
```

Comment:
Right now there should not be an aggregated item left on the buffer. We dequeue them right away. Maybe we need to change this to Signed only.

Reply:
yeah I realized this too